python/cuml/preprocessing/model_selection.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf
+from typing import Union, Tuple
+from numba import jit
+import numpy as np
+
+
+@jit
+def _shuffle_idx(df: cudf.DataFrame) -> np.ndarray:
+    """ Return indices which represent a randomly shuffled version of `df`
+    """
+    return np.random.permutation(len(df))
+
+
+def train_test_split(
+    X: cudf.DataFrame,
+    y: Union[str, cudf.Series],
+    train_size: Union[float, int] = 0.8,
+    shuffle: bool = True,
+    seed: int = None,
+) -> Tuple[cudf.DataFrame, cudf.DataFrame, cudf.DataFrame, cudf.DataFrame]:
+    """
+    Partitions the data into four collated dataframes, mimicing sklearn's
+    `train_test_split`
+
+    Parameters
+    ----------
+    X : cudf.DataFrame
```

Comment:
Will we want to support non cuDF input types like the rest of cuML (particularly after PR 612)? We then can use cupy's indexing instead to support this for all inputs instead. I'd be fine with making that a todo for being done after a merge of 612.

Reply:
I had it like this because it was the only way I could find to have consistent, reasonably efficient indexing. If cupy indexing can support this, and we're okay with adding in the dependency, then I'll put in a todo for it